README
Devel::Pragma version 1.0.1
================================

This module provides helper functions for developers of lexical pragmas, though a few functions may
be useful to non-pragma developers as well.

Pragmas can be used both in older versions of perl (from 5.8.1), which had limited support, and in
the most recent versions, which have improved support.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

    Lexical::SealRequireHints

COPYRIGHT AND LICENCE

Copyright (C) 2008-2016 by chocolateboy

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Devel::Pragma.

1.0.1 Tue 28 Jun 10:09:48 2016
    - META.yml: add the bugtracker link
    - META.yml: fix the version

1.0.0 Tue 28 Jun 09:15:13 2016
    - remove on_require

0.62  Mon 26 Oct 12:13:17 2015
    - update ppport.h
    - update copyright year

0.61  Mon 26 Oct 11:33:19 2015
    - specify minimum perl version (thanks Lisa Hare)

0.60  Sun Oct 20 12:24:09 2013
    - use Lexical::SealRequireHints to fix %^H bugs
    - rename my_hints -> hints (my_hints is still supported, but deprecated)
    - update ppport.h from 3.19 to 3.20
    - code cleanup

0.54  Mon Aug 16 13:54:36 2010
    - portability fixes for Visual Studio and older GCCs

0.53  Sun Feb 22 01:44:51 2009
    - yet another metadata fix

0.52  Thu Feb 19 19:47:20 2009
    - refix META.yml

0.51  Thu Feb 19 16:38:56 2009
    - fix META.yml

0.50  Thu Feb 19 03:56:55 2009
    - added on_require
        - added require.t
    - ported to B::Hooks::OP::Check/B::Hooks::OP::Annotation

0.40  Sun Feb 1 19:36:21 2009
    - portability fixes
    - added new function: fqname
        - added fqname.t
    - made require hook exception-safe (Marcus Holland-Moritz++)
        - added exception.t

0.32  Fri Jan 9 15:26:21 2009
    - added scope() - unique identifier for the currently-compiling scope
        - added scope.t
    - set $^H bit on import
    - fix :all tag
        - added tag.t
    - fix compile-time lexical scoping (leak.t)

0.31  Tue Jan 6 19:55:08 2009
    - add missing dependency on B::Hooks::EndOfScope (which bumps the minimum perl version up to 5.8)

0.30  Tue Jan 6 19:03:06 2009
    - more robust/less disruptive %^H clearance
    - code cleanup
    - make my_hints apply #33311 lexically (again)

0.22  Tue Jan 6 10:19:40 2009
    - refine exclusions

0.21  Tue Jan 6 08:10:45 2009
    - Fix optimize flags in Makefile.PL

0.20  Tue Jan 6 07:16:42 2009
    - apply the %^H leak fix (by patching PL_ppaddr[OP_REQUIRE] and PL_ppaddr[OP_DOFILE])
      globally rather than lexically
    - added leak.t to confirm the bug on perls prior to patch #33311 if Devel::Pragma is not used
    - added new_scope.t
    - documentation tweaks

0.12  Mon Jun 23 23:54:04 2008
    - added an optional argument to new_scope to distinguish
      between different classes that share the same code

0.11  Mon Jun 23 21:03:44 2008
    - renamed Devel::Pragma
    - added function: ccstash
        - added ccstash.t
    - fixed MANIFEST (thanks Taro Nishino)

0.10  Mon Jun 23 14:40:42 2008
    - fix bug that was causing my_hints to clobber %^H
    - added new_scope function
    - removed my_hints aliases

0.09  Sat Jun 21 11:06:31 2008
    - bugfix: return reference to %^H when already in scope

0.08  Sat Jun 21 00:05:44 2008
    - doc fixlet

0.07  Fri Jun 20 23:47:28 2008
    - doc fixlet

0.06  Fri Jun 20 20:35:45 2008
    - my_hh returns a reference to %^H
    - added another alias: my_hints
    - future-harden: use the outermost unused PL_hints bit

0.05  Tue Jun 17 23:40:36 2008
    - lexicalize_hh is now an alias for my_hh

0.04  Tue Jun 17 23:07:09 2008
    - added dependency on if.pm for perl 5.6.0 and 5.6.1

0.03  Tue Jun 17 21:28:10 2008
    - added dependency on Test::More for perl 5.6.0 and 5.6.1

0.02  Tue Jun 17 21:15:55 2008
    - move minimum perl version down from 5.8.0 to 5.6.0 (only tested on 5.6.2)

0.01  Tue Jun 17 20:14:34 2008
    - original version; created by h2xs 1.23 with options -A -c -n Devel::Hints::Lexical


t/test_1.pm
package test_1;

use strict;
use warnings;

our $test;

# 0 for failure if the key has leaked
# 1 for success if it hasn't
BEGIN { $test = exists($^H{'Devel::Pragma::Test'}) ? 0 : 1 }

sub test() { $test }

1;


t/test_5.pm
package test_5;

use strict;
use warnings;

our $test;

# 0 for failure if the key has leaked
# 1 for success if it hasn't
BEGIN { $test = exists($^H{'Devel::Pragma::Test'}) ? 0 : 1 }

sub test() { $test }

1;


t/test_12.pm
package test_12;

use strict;
use warnings;

our $test;

# 0 for failure if the key has leaked
# 1 for success if it hasn't
BEGIN { $test = exists($^H{'Devel::Pragma::Test'}) ? 0 : 1 }

sub test() { $test }

1;


Makefile.PL
use 5.008001;

use strict;
use warnings;

use ExtUtils::MakeMaker;
use Config;

my $EUMM_VERSION = eval($ExtUtils::MakeMaker::VERSION);
my $OPTIMIZE;

if ($Config{gccversion}) {
    $OPTIMIZE = '-O3 -Wall';
    $OPTIMIZE .= ' -g -Wextra -Wdeclaration-after-statement' if (-d 'dev');
} elsif ($Config{osname} eq 'MSWin32') {
    $OPTIMIZE = '-O2 -W4';
} else {
    $OPTIMIZE = $Config{optimize};
}

WriteMakefile(
    NAME          => 'Devel::Pragma',
    VERSION_FROM  => 'lib/Devel/Pragma.pm',
    PREREQ_PM     => {
        'Lexical::SealRequireHints' => '0.007',
    },
    ABSTRACT_FROM => 'lib/Devel/Pragma.pm',
    AUTHOR        => 'chocolateboy <chocolate@cpan.org>',
    LIBS          => [''],
    DEFINE        => '',
    INC           => '-I.',
    OPTIMIZE      => $OPTIMIZE,
    ($EUMM_VERSION >= 6.31 ? (LICENSE => 'perl') : ()),
    ($EUMM_VERSION >= 6.48 ? (MIN_PERL_VERSION => '5.8.1') : ()),
    ($EUMM_VERSION >= 6.46 ?
        (META_MERGE => {
            resources => {
                repository => 'http://github.com/chocolateboy/Devel-Pragma',
            },
        })
        : ()
    ),
);


lib/Devel/Pragma.pm
package Devel::Pragma;

use 5.008001;

# make sure this is loaded first
use Lexical::SealRequireHints;

use strict;
use warnings;

use Carp qw(carp croak);
use Scalar::Util;
use XSLoader;

use base qw(Exporter);

our $VERSION = '1.0.1';
our @EXPORT_OK = qw(my_hints hints new_scope ccstash scope fqname);
our %EXPORT_TAGS = (all => [ @EXPORT_OK ]);

XSLoader::load(__PACKAGE__, $VERSION);

# return a reference to the hints hash
sub my_hints() {
    # set HINT_LOCALIZE_HH (0x20000)
    $^H |= 0x20000;
    return \%^H;
}

BEGIN { *hints = \&my_hints }

# make sure the "enable lexically-scoped %^H" flag is set (on by default in 5.10)
sub check_hints() {
    unless ($^H & 0x20000) {
        carp('Devel::Pragma: unexpected $^H (HINT_LOCALIZE_HH bit not set) - setting it now, but results may be unreliable');
    }
    return hints; # create it if it doesn't exist - in some perls, it starts out NULL
}

# return a unique integer ID for the current scope
sub scope() {
    check_hints;
    xs_scope();
}

# return a boolean indicating whether this is the first time "use MyPragma" has been called in this scope
sub new_scope(;$) {
    my $caller = shift || caller;
    my $hints = check_hints();

    # this is %^H as an integer - it changes as scopes are entered/exited i.e. it's a unique
    # identifier for the currently-compiling scope (the scope in which new_scope
    # is called)
    #
    # we don't need to stack/unstack it in %^H as %^H itself takes care of that
    # note: we need to call this *after* %^H is referenced (and possibly autovivified) above
    #
    # every time new_scope is called, we write this scope ID to $^H{"Devel::Pragma::new_scope::$caller"}.
    # if $^H{"Devel::Pragma::new_scope::$caller"} == scope() (i.e. the stored scope ID is the same as the
    # current scope ID), then we're augmenting the current scope; otherwise we're in a new scope - i.e.
    # a nested or outer scope that didn't previously "use MyPragma"

    my $current_scope = scope();
    my $id = "Devel::Pragma::new_scope::$caller";
    my $old_scope = exists($hints->{$id}) ? $hints->{$id} : 0;
    my $new_scope; # is this a scope in which new_scope has not previously been called?

    if ($current_scope == $old_scope) {
        $new_scope = 0;
    } else {
        $hints->{$id} = $current_scope;
        $new_scope = 1;
    }

    return $new_scope;
}

# given a short name (e.g. "foo"), expand it into a fully-qualified name with the caller's package prefixed
# e.g. "main::foo"
#
# if the name is already fully-qualified, return it unchanged
sub fqname ($;$) {
    my $name = shift;
    my ($package, $subname);

    $name =~ s{'}{::}g;

    if ($name =~ /::/) {
        ($package, $subname) = $name =~ m{^(.+)::(\w+)$};
    } else {
        my $caller = @_ ? shift : ccstash();
        ($package, $subname) = ($caller, $name);
    }

    return wantarray ? ($package, $subname) : "$package\::$subname";
}

# helper function: return true if $ref ISA $class - works with non-references, unblessed references and objects
sub _isa($$) {
    my ($ref, $class) = @_;
    return Scalar::Util::blessed($ref) ? $ref->isa($class) : ref($ref) eq $class;
}

# make sure "enable lexically-scoped %^H" is set in older perls, and export the requested functions
sub import {
    my $class = shift;
    $^H |= 0x20000; # set HINT_LOCALIZE_HH (0x20000)
    $class->export_to_level(1, undef, @_);
}

1;

__END__

=head1 NAME

Devel::Pragma - helper functions for developers of lexical pragmas

=head1 SYNOPSIS

    package MyPragma;

    use Devel::Pragma qw(:all);

    sub import {
        my ($class, %options) = @_;
        my $hints  = hints;        # lexically-scoped %^H
        my $caller = ccstash();    # currently-compiling stash

        unless ($hints->{MyPragma}) { # top-level
            $hints->{MyPragma} = 1;
        }

        if (new_scope($class)) {
            ...
        }

        my $scope_id = scope();
    }

=head1 DESCRIPTION

This module provides helper functions for developers of lexical pragmas, though a few functions may
be useful to non-pragma developers as well.

Pragmas can be used both in older versions of perl (from 5.8.1), which had limited support, and in
the most recent versions, which have improved support.

=head1 EXPORTS

C<Devel::Pragma> exports the following functions on demand. They can all be imported at once by using
the C<:all> tag. e.g.

    use Devel::Pragma qw(:all);

=head2 hints

This function enables the scoped behaviour of the hints hash (C<%^H>) and then returns a reference to it.

The hints hash is a compile-time global variable (which is also available at runtime in recent perls) that
can be used to implement lexically-scoped features and pragmas. This function provides a convenient
way to access this hash without the need to perform the bit-twiddling that enables it on older perls.
In addition, this module loads L<Lexical::SealRequireHints>, which implements bugfixes
that are required for the correct operation of the hints hash on older perls (< 5.12.0).

Typically, C<hints> should be called from a pragma's C<import> (and optionally C<unimport>) method:

    package MyPragma;

    use Devel::Pragma qw(hints);

    sub import {
        my $class = shift;
        my $hints = hints;

        if ($hints->{MyPragma}) {
            # ...
        } else {
            $hints->{MyPragma} = ...;
        }

        # ...
    }

=head2 new_scope

This function returns true if the currently-compiling scope differs from the scope being compiled the last
time C<new_scope> was called. Subsequent calls will return false while the same scope is being compiled.

C<new_scope> takes an optional parameter that is used to uniquely identify its caller. This should usually be
supplied as the pragma's class name unless C<new_scope> is called by a module that is not intended
to be subclassed. e.g.

    package MyPragma;

    sub import {
        my ($class, %options) = @_;

        if (new_scope($class)) {
            ...
        }
    }

If not supplied, the identifier defaults to the name of the calling package.

=head2 scope

This returns an integer that uniquely identifies the currently-compiling scope. It can be used to
distinguish or compare scopes.

A warning is issued if C<scope> (or C<new_scope>) is called in a context in which it doesn't make sense i.e. if the
scoped behaviour of C<%^H> has not been enabled - either by explicitly modifying C<$^H>, or by calling
L<"hints">.

=head2 ccstash

This returns the name of the currently-compiling stash. It can be used as a replacement for the scalar form of
C<caller> to provide the name of the package in which C<use MyPragma> is called. Unlike C<caller>, it
returns the same value regardless of the number of intervening calls before C<MyPragma::import>
is reached.

e.g. given a pragma:

    package MySuperPragma;

    use Devel::Hints qw(ccstash);

    sub import {
        my ($class, %options) = @_;
        my $caller = ccstash();

        no strict 'refs';

        *{"$caller\::whatever"} = ... ;
    }

and a subclass:

    package MySubPragma

    use base qw(MySuperPragma);

    sub import {
        my ($class, %options) = @_;
        $class->SUPER::import(...);
    }

and a script that uses the subclass:

    #!/usr/bin/env perl

    use MySubPragma;

- the C<ccstash> call in C<MySuperPragma::import> returns the name of the package that's being compiled when
the call to C<MySuperPragma::import> (via C<MySubPragma::import>) takes place i.e. C<main> in this case.

=head2 fqname

Given a subroutine name, usually supplied by the caller of the pragma's import method, this function returns
the name in package-qualified form. In addition, old-style C<'> separators are converted to new-style C<::>.

If the name contains no separators, then the optional calling package is prepended. If not supplied, the caller
defaults to the value returned by L<"ccstash">. If the name is already package-qualified,
then it is returned unchanged.

In list context, C<fqname> returns the package and unqualified subroutine name (e.g. 'main' and 'foo'), and in scalar
context it returns the package and sub name joined by '::' (e.g. 'main::foo').

e.g.

    package MyPragma;

    sub import {
        my ($class, @names) = @_;

        for my $name (@names) {
            my $fqname = fqname($name);
            say $fqname;
        }
    }

    package MySubPragma;

    use base qw(MyPragma);

    sub import { shift->SUPER::import(@_) }

    #!/usr/bin/env perl

    use MyPragma qw(foo Foo::Bar::baz Foo'Bar'baz Foo'Bar::baz);

    {
        package Some::Other::Package;

        use MySubPragma qw(quux);
    }

prints:

    main::foo
    Foo::Bar::baz
    Foo::Bar::baz
    Foo::Bar::baz
    Some::Other::Package::quux

=head1 VERSION

1.0.1

=head1 SEE ALSO

=over

=item * L<pragma|pragma>

=item * L<perlpragma|perlpragma>

=item * L<perlvar|perlvar>

=item * L<B::Hooks::EndOfScope|B::Hooks::EndOfScope>

=item * L<B::Hooks::OP::Check|B::Hooks::OP::Check>

=item * L<B::Hooks::OP::PPAddr|B::Hooks::OP::PPAddr>

=item * L<B::Hooks::OP::Annotation|B::Hooks::OP::Annotation>

=item * L<Devel::Hints|Devel::Hints>

=item * L<Lexical::SealRequireHints|Lexical::SealRequireHints>

=item * http://tinyurl.com/45pwzo

=back

=head1 AUTHOR

chocolateboy <chocolate@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2016 by chocolateboy

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.1 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Devel/Hints/Lexical.pm
package Devel::Hints::Lexical;

use 5.006;

use strict;
use warnings;

our $VERSION = '0.10';

use XSLoader;
use Scope::Guard;

use base qw(Exporter);

our @EXPORT_OK = qw(my_hints new_scope);

XSLoader::load(__PACKAGE__, $VERSION);

sub new_scope() {
    # this is %^H as an integer - it changes as scopes are entered/exited i.e. it's a unique
    # identifier for the currently compiling scope (the scope in which new_scope 
    # is called)
    #
    # we don't need to stack/unstack it in %^H as %^H itself takes care of that
    # note: we need to call this *after* %^H is referenced (and possibly autovivified) by my_hints
    #
    # every time new_scope is called, we write this scope ID to $^H{"Devel::Hints::Lexical::Scope::$caller"}.
    # if $^H{"Devel::Hints::Lexical::Scope::$caller"} == _scope() (i.e. the stored scope ID is the same as the
    # current scope ID), then we're augmenting the current scope; otherwise we're in a new scope - i.e.
    # a nested or outer scope that didn't previously "use MyPragma"

    # set HINT_LOCALIZE_HH (0x20000) + an unused bit (0x80000000) so that
    # this module (which can't use itself) can work around the %^H bug

    $^H |= 0x80020000;
    $^H{'Devel::Hints::Lexical'} = 1;

    my $current_scope = _scope();
    my $key = 'Devel::Hints::Lexical::Scope::' . caller;
    my $old_scope = exists($^H{$key}) ? $^H{$key} : 0;
    my $new_scope; # is this a scope in which new_scope has not previously been called?

    if ($current_scope == $old_scope) {
        $new_scope = 0;
    } else {
        $^H{$key} = $current_scope;
        $new_scope = 1;
    }

    return $new_scope;
}

sub my_hints() {
    if (new_scope) {
        my $guard = Scope::Guard->new(\&_leave);

        $^H{$guard} = $guard;

        _enter();
    }

    return \%^H;
}

1;

__END__

=head1 NAME

Devel::Hints::Lexical - lexical pragma utils

=head1 SYNOPSIS

  package MyPragma;

  use Devel::Hints::Lexical qw(my_hints);

  sub import {
      my $hints = my_hints;
      $hints->{MyPragma} = 1;
  }

=head1 DESCRIPTION

This module provides helper functions for developers of lexical pragmas. These can be used both in older versions of
perl (from 5.6.0), which have limited support for lexical pragmas, and in the most recent versions, which have improved
support.

=head1 EXPORTS

=head2 my_hints

Until perl change #33311, which isn't currently available in any stable perl release, %^H is dynamically-scoped,
rather than lexically-scoped. This means that values set in %^H are visible in modules loaded by C<use>.
This makes pragmas leak from the scope in which they're meant to be enabled into scopes in which they're
not. C<my_hints> fixes that by making %^H lexically scoped i.e. it prevents %^H leaking across file boundaries.

C<my_hints> installs versions of perl's C<require> and C<do EXPR> builtins in the
currently-compiling scope which clear %^H before they execute and restore the previous %^H afterwards.
Thus it can be thought of a lexically-scoped backport of change #33311.

Note that C<my_hints> also sets the $^H bit that "localizes" (or in this case "lexicalizes") %^H.

The return value is a reference to %^H.

=head2 new_scope

This function returns true if the currently-compiling scope differs from the scope being compiled the last
time C<new_scope> was called. Subsequent calls will return false while the same scope is being compiled.

=head1 VERSION

0.10

=head1 SEE ALSO

=over

=item * L<perlpragma|perlpragma>

=item * L<perlvar|perlvar>

=item * L<Devel::Hints|Devel::Hints>

=item * http://tinyurl.com/45pwzo

=back

=head1 AUTHOR

chocolateboy <chocolate.boy@email.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by chocolateboy

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
